File: clojure/anagram/src/anagram.clj
(ns anagram
  (:require [clojure.string :as str]))

(defn- anagram? [subject candidate]
  (let [w0 (str/lower-case subject)
        w1 (str/lower-case candidate)]
    (and (not= w0 w1)
         (= (sort w0) (sort w1)))))

(defn anagrams-for [subject candidates]
  (filter #(anagram? subject %) candidates))

File: elixir/anagram/lib/anagram.ex
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t(), [String.t()]) :: [String.t()]
  def match(base, candidates) do
    candidates
    |> reject_candidate_equal(base)
    |> filter_candidate_with_same_characters_than(base)
  end

  defp reject_candidate_equal(candidates, base) do
    candidates
    |> Enum.reject(& String.downcase(&1) == String.downcase(base))
  end

  defp filter_candidate_with_same_characters_than(candidates, base) do
    candidates
    |> Enum.filter(& format(&1) == format(base))
  end

  defp format(word), do: String.downcase(word) |> String.codepoints |> Enum.sort |> Enum.join
end
